### PR TITLE
Refactor event tracking infrastructure

### PR DESCRIPTION
**Pull request by aaemnnosttv**
## Summary

Addresses issue #1055 

## Relevant technical choices
* Uses a custom `dataLayer` global  
See https://developers.google.com/gtagjs/devguide/rename

## Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 4.7 and PHP 5.6.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

### CODE REVIEW

File: assets/js/googlesitekit-adminbar-loader.js [hunk 1]
```diff
@@ -23,9 +23,9 @@
  */
 import {
 	appendNotificationsCount,
-	sendAnalyticsTrackingEvent,
 	getQueryParameter,
 } from './util/standalone';
+import { trackEvent } from './util/tracking';
```
[the remark this answers is not in the document]
Comment (aaemnnosttv) [reply assets/js/googlesitekit-adminbar-loader.js]: Are you referring to using `GoogleUtil` instead? I could be mistaken but I thought we were maybe planning on moving away from using those and using relative imports for everything?

File: assets/js/util/tracking/index.js [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+		userIDHash,
+	} = _global._googlesitekitBase || DEFAULT_CONFIG;
+
+	Object.assign( config, DEFAULT_CONFIG, {
+		isFirstAdmin,
+		trackingID,
+		trackingEnabled,
+		userIDHash,
+		// Remove any trailing slash from the reference URL.
+		referenceSiteURL: referenceSiteURL.toString().replace( /\/+$/, '' ),
```
[the remark this answers is not in the document]
Comment (aaemnnosttv) [reply assets/js/util/tracking/index.js]: Yeah, it's not probably needed anymore since we updated the reference URL to always be without a trailing slash in WP, but this does pass through the `googlesitekit_inline_base_data` after that so it's technically possible it could have changed. Basically inconsistencies in the (same) URLs would probably throw a wrench in reporting.

File: assets/js/util/tracking/index.js [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+	if ( enableTrackingPromise ) {
+		return enableTrackingPromise;
+	}
+
+	config.trackingEnabled = true;
+
+	dataLayerPush( _global, 'js', new Date() );
+	dataLayerPush( _global, 'config', config.trackingID );
+
+	// If the script is already in the DOM then we shouldn't get here as the promise should already be returned.
```
[the remark this answers is not in the document]
Comment (aaemnnosttv) [reply assets/js/util/tracking/index.js]: Good question; I think that was left from earlier in the lifecycle of the branch. Really the only way it should exist on the page now is if tracking were to be bootstrapped again while calling for the promise to be cleared. I think perhaps I thought we might cover this in the tests but obv. that's not happening (yet). It's definitely an edge case; not sure if it warrants an error logged. Maybe a warning?

File: assets/js/util/tracking/index.test.js [hunk 4]
```diff
@@ -0,0 +1,41 @@
+/**
+ * Internal dependencies
+ */
+import { bootstrapTracking, DATA_LAYER, trackEvent } from './';
```
[the remark this answers is not in the document]
Comment (aaemnnosttv) [reply assets/js/util/tracking/index.test.js]: > I wonder: would it make sense to use react-helmet for this instead of our document.append? It's a bit of a nicer API and could make things easier to test, but I don't have strong feelings on this one as we don't have much of a need for it otherwise from what I can tell right now.

The ACs say that the module shouldn't depend on any third-party code.

+1 for this test though. We could mock `document.head.appendChild` but I would rather just check that the element is there 👍

File: assets/js/googlesitekit-adminbar-loader.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (tofumatt) [reply assets/js/googlesitekit-adminbar-loader.js]: Yes, and I think the best way going forward is to use project-absolute imports (eg `assets/js/[...]` going forward).

Relative imports are a pain to refactor and deal with in my experience when moving around files or creating new files. So `GoogleUtil/tracking` is an odd path, but we can easily change it to `assets/js/util/tracking` or whatever when we make that change en-masse. `./util/tracking` is a lot tougher to find+replace.

File: assets/js/util/tracking/index.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (tofumatt) [reply assets/js/util/tracking/index.js]: Would be worth leaving a comment there explaining that, but sounds good 👍

File: assets/js/util/tracking/index.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (tofumatt) [reply assets/js/util/tracking/index.js]: Yeah, if we encounter this it'll still work, but worth logging a warning saying something went a bit wonky 👍

File: assets/js/util/tracking/index.test.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (tofumatt) [reply assets/js/util/tracking/index.test.js]: Ah, right on. I think this is fine anyway and helmet would be overkill.

Usually with testing-library it renders stuff into a `<div>` and not a full-page… so we may have to test the call to `appendChild`, but I think it's worth exploring this in a separate issue. (#1083 maybe.)

File: assets/js/util/tracking/index.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (felixarntz) [reply assets/js/util/tracking/index.js]: Yeah, I think we should continue to ensure that the slash is stripped here.

File: assets/js/util/tracking/index.js [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+}
+
+/**
+ * Change the active state of tracking.
+ *
+ * @param {boolean} activeStatus The new state to set.
+ * @param {Object} _global The global object. Optional. (Used for testing only)
+ * @return {Promise} A promise that resolves with an object { trackingEnabled: (bool) }.
+ */
+export async function toggleTracking( activeStatus, _global = global ) {
```
Comment (felixarntz) [inline assets/js/util/tracking/index.js]: While the dependency injection of `global` seems neat, I don't like having that in a "public" function like here. _If_ we really need this, then let's assign and re-use the `_global` set in `bootstrapTracking()` and remove the parameter from the other functions.

However, I'm also curious: What does passing `_global` actually allows us to do that we wouldn't be able to do otherwise? Why not just access `global`?

File: assets/js/util/tracking/index.js [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+ */
+export const DATA_LAYER = '_googlesitekitDataLayer';
+
+/**
+ * Initializes tracking.
+ *
+ * @param {Object} _global The global object. Optional. (Used for testing only)
+ * @param {boolean} _resetPromise Clears the internal reference to the enableTrackingPromise (Optional - for testing only)
+ */
+export async function bootstrapTracking( _global = global, _resetPromise = false ) {
```
Comment (felixarntz) [inline assets/js/util/tracking/index.js]: If we need flexibility here, I'd much rather have us pass `initialConfig` as an object and `dataLayer` as an array. The latter could be stored for future access. Let's be specific rather than passing the entire `_global`.

File: assets/js/util/tracking/index.js [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+	if ( enableTrackingPromise ) {
+		return enableTrackingPromise;
+	}
+
+	config.trackingEnabled = true;
+
+	dataLayerPush( _global, 'js', new Date() );
+	dataLayerPush( _global, 'config', config.trackingID );
+
+	const { document } = _global;
```
Comment (felixarntz) [inline assets/js/util/tracking/index.js]: This should really come from `global`. We don't need to go as far as support custom mocks for `document`.

File: assets/js/util/tracking/index.js [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+
+/**
+ * gtag script identifier.
+ */
+const SCRIPT_IDENTIFIER = 'data-googlesitekit-gtag';
+
+/**
+ * Data layer global used for internal/private Site Kit data.
+ */
+export const DATA_LAYER = '_googlesitekitDataLayer';
```
Comment (felixarntz) [inline assets/js/util/tracking/index.js]: That's really neat that we can use our own global and pass it - will help with interoperability and scoping our own data.

File: includes/Core/Util/Tracking.php [hunk 9]
```diff
@@ -198,7 +118,8 @@ private function print_standalone_event_tracking_script() {
 	 * @return array Filtered $data.
 	 */
 	private function inline_js_admin_data( $data ) {
```
Comment (felixarntz) [inline includes/Core/Util/Tracking.php]: Let's rename this to `inline_js_base_data()` now that it's no longer attached to the admin data filter.

File: assets/js/util/tracking/index.js (hunk 7 above)
Comment (aaemnnosttv) [reply assets/js/util/tracking/index.js]: See @tofumatt 's comment here https://github.com/google/site-kit-wp/pull/1079#discussion_r371776734

Comment (felixarntz) [reply assets/js/util/tracking/index.js]: Okay, that's fair. One thing we shouldn't do though (as mentioned in https://github.com/google/site-kit-wp/pull/1079#discussion_r373583361) is not add this kind of arguments in public or high-level API functions. This is very low-level functionality and should be handled in a low-level API.

Although we don't currently plan exposing these functions, they are still the "public" functions of this module and only should include minimum parameters that are needed for the primary purpose.

File: assets/js/util/tracking/index.js (hunk 6 above)
Comment (felixarntz) [inline assets/js/util/tracking/index.js]: This function as well as related constants should be moved to an `index.private.js` file. This is our convention previously established for anything that needs to be exported, but _only_ needs to be exported for internal things like testability.

That file should for example also include the `config` variable. That variable could also be exported, and then imported by `index.js` to access it.

File: assets/js/util/tracking/index.js (hunk 5 above)
Comment (aaemnnosttv) [reply assets/js/util/tracking/index.js]: I don't think it offers additional capabilities, but is more of a pattern to make the usage more explicit by avoiding reaching outside of the function from within. That is, it's a bit closer to writing pure functions. I'm not sure if global contamination is a concern with Jest or not, but this avoids state leaking between tests by using a dedicated object instead of the real `global`.

To your point though, we can easily scope this to the module rather than each function so I'll update that. We're already doing that with the `enableTrackingPromise` and `config` so I don't see why we can't do that with `global` and avoid the extra parameter to each function.

**Comment by aaemnnosttv** [comment]
@felixarntz I've simplified the implementation quite a bit. The async behavior was unnecessary and causing problems with the tests since the injected script tag's `onload` callback was never called. I also simplified many of the functions which had unnecessary return values. The only important return value is from `isTrackingEnabled`.

Ready for another pass 👍

File: assets/js/util/tracking/index.js [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+ * Internal dependencies
+ */
+import { bootstrapTracking } from './index.private';
+export {
+	disableTracking,
+	enableTracking,
+	isTrackingEnabled,
+	toggleTracking,
+	trackEvent,
+} from './index.private';
```
Comment (felixarntz) [inline assets/js/util/tracking/index.js]: Can we instead declare the public interface in this file, similar like we're doing in the other new modules (e.g. `googlesitekit/api` and `googlesitekit/api/cache`)? This here works from an interface standpoint, but these functions are the public part of the module and therefore shouldn't be declared in the `.private` file.

We'll need to export some more variables from `index.private.js` then, but that'd be alright.
